src/query/storage/m3/cluster_resolver.go

```diff
@@ -0,0 +1,299 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package m3
+
+import (
+	"sort"
+	"time"
+
+	"github.com/m3db/m3/src/query/storage"
+)
+
+type unaggregatedNamespaceType uint8
+
+const (
+	partiallySatisfiesRange unaggregatedNamespaceType = iota
+	fullySatisfiesRange
+	disabled
+)
+
+type unaggregatedNamespaceDetails struct {
+	satisfies        unaggregatedNamespaceType
+	clusterNamespace ClusterNamespace
+	retention        time.Duration
+}
+
+// resolveUnaggregatedNamespaceForQuery determines if the unaggregated namespace
+// should be used, and if so, determines if it fully satisfies the query range.
+func resolveUnaggregatedNamespaceForQuery(
+	now time.Time,
+	start time.Time,
+	unaggregated ClusterNamespace,
+	opts *storage.FanoutOptions,
+) unaggregatedNamespaceDetails {
+	if opts.FanoutUnaggregated == storage.FanoutForceDisable {
+		return unaggregatedNamespaceDetails{satisfies: disabled}
+	}
+
+	retention := unaggregated.Options().Attributes().Retention
+	unaggregatedStart := now.Add(-1 * retention)
+	if unaggregatedStart.Before(start) || unaggregatedStart.Equal(start) {
+		return unaggregatedNamespaceDetails{
+			clusterNamespace: unaggregated,
+			satisfies:        fullySatisfiesRange,
```

Comment:
nit: Is it worth including `retention` in return for this code path?  In case it's needed in future and we don't find out that it's set to zero with tests or something?

Reply:
Sure, sounds good.